### PR TITLE
Add boolean indexing function to jax.ops

### PR DESCRIPTION
**Pull request by ibab**
We sometimes need to perform boolean indexing inside of `jax.jit`, and found that it's useful to have a function which does that while padding the output to the same shape as the original input.
I think that having this function will be useful, even if JAX gains better variable shape support in the future.

Some things I'm not sure about in this implementation:
 - Is `jax.lax.sort_key_val` going to do a stable sort? XLA exposes a parameter for setting the sort to stable, but this is not exposed in `jax.lax.sort*`. I'd be happy to add an additional argument or to default the sort to be stable.
 - I can get the code to throw an error saying "This is a bug in JAX's shape-checking rules; please report it!". How can I perform proper shape checking, or should I perhaps add the checks to `jax.sort*`?

### CODE REVIEW

**Comment by shoyer** [comment]
I think the shape checking needs to happen in `lax.sort_key_val`, before calling the `sort_key_val_p`. I'm guessing that's also where you would add the `stable` parameter, which might be as simple as passing a boolean as a keyword (but not I'm not quite confident enough about XLA-Python to tell you for sure).

File: jax/ops/scatter.py [hunk 1]
```diff
@@ -267,3 +270,19 @@ def segment_sum(data, segment_ids, num_segments=None):
   out = np.zeros((num_segments,) + data.shape[1:], dtype=data.dtype)
   segment_ids = np.mod(segment_ids, num_segments)
   return index_add(out, segment_ids, data)
+
+
+def index_by_mask(x, boolean_mask, axis=0, pad_value=0):
+  """A version of NumPy's `x[boolean_mask]`, but with a static output shape.
```
Comment (hawkinsp) [inline jax/ops/scatter.py:276]: This is restructured text, not markdown, and I think the backticks need to be doubled or written something like:
```
:code:`x[boolean_mask]`
```

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Sorry for the slow review.

A heads up: we plumbed the sort stability option into `lax`. So your sort should now be stable. Sync up?

Otherwise I'm happy merging this.

**Comment by cghawthorne** [comment]
I recently found this pattern very helpful, but I only knew about it because I happened to see this PR. Would it be possible to get this "official" implementation checked in so the pattern can be more easily used? If there's anything I can do toward that, I'm happy to help.

**Comment by jakevdp** [comment]
I'm cleaning up some old PRs – I think it's an interesting idea to add a function that's something like this, but the JAX package has evolved so much in the meantime that if we do it it probably makes sense to open a new PR. Thanks!
